File: canConstruct.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function (ransomNote, magazine) {
  let hashmap = new Map();
  for (const v of magazine) {
    let c = hashmap.get(v) + 1 || 1;
    hashmap.set(v, c);
  }
  for (const v of ransomNote) {
    let c = hashmap.get(v);
    if (!c) return false;
    hashmap.set(v, c - 1);
  }
  return true;
};

File: leetcode/canConstruct.py
from collections import Counter


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        return can_construct(ransomNote, magazine)


def can_construct(ransomNote, magazine):
    c_m = Counter(magazine)
    for c in ransomNote:
        if c not in c_m or c_m[c] == 0:
            return False
        c_m[c] -= 1
    return True
